File: backend/app/core/state_machines.py
"""State machines explícitas de dominio (ARCHITECTURE.md Pattern 5).

Módulo PURO: no importa ORM ni FastAPI — solo los enums de estado. Phase 5/6
importan ``validar_transicion()`` en sus services y traducen
``TransicionInvalidaError`` a 409 Conflict en el router (el dominio no decide
códigos HTTP). Los tests unitarios de este módulo corren en milisegundos sin
levantar la base ni el stack.

Cada ``*_TRANSITIONS`` dict mapea ``estado_origen -> {estados_destino}``.
Un ``set()`` vacío marca un estado TERMINAL (no admite más transiciones).
"""

from app.models.mesa import EstadoMesa
from app.models.pedido import EstadoPedido
from app.models.reserva import EstadoReserva
from app.models.sesion_mesa import EstadoSesion
from app.models.pago import EstadoPago


class TransicionInvalidaError(Exception):
    """Transición de estado no permitida por la máquina de estados del dominio.

    Atributos:
        maquina: nombre de la máquina ("mesa", "pedido", ...).
        actual: estado origen (Enum).
        nueva: estado destino rechazado (Enum).
    """

    def __init__(self, maquina: str, actual, nueva):
        self.maquina = maquina
        self.actual = actual
        self.nueva = nueva
        super().__init__(
            f"[{maquina}] transición {actual!r} → {nueva!r} no permitida"
        )


# ARCHITECTURE.md Pattern 5 — set vacío = terminal.
MESA_TRANSITIONS: dict[EstadoMesa, set[EstadoMesa]] = {
    EstadoMesa.disponible: {EstadoMesa.reservada, EstadoMesa.ocupada},
    EstadoMesa.reservada: {EstadoMesa.ocupada, EstadoMesa.disponible},
    EstadoMesa.ocupada: {EstadoMesa.limpieza},
    EstadoMesa.limpieza: {EstadoMesa.disponible},
}

PEDIDO_TRANSITIONS: dict[EstadoPedido, set[EstadoPedido]] = {
    EstadoPedido.borrador: {EstadoPedido.enviado},
    EstadoPedido.enviado: {EstadoPedido.aceptado, EstadoPedido.rechazado},
    EstadoPedido.aceptado: {EstadoPedido.en_preparacion},
    EstadoPedido.en_preparacion: {EstadoPedido.servido},
    EstadoPedido.servido: {EstadoPedido.pagado},
    EstadoPedido.rechazado: set(),  # terminal
    EstadoPedido.pagado: set(),  # terminal
}

RESERVA_TRANSITIONS: dict[EstadoReserva, set[EstadoReserva]] = {
    EstadoReserva.pendiente: {EstadoReserva.confirmada, EstadoReserva.cancelada},
    EstadoReserva.confirmada: {EstadoReserva.cancelada},
    EstadoReserva.cancelada: set(),  # terminal
}

PAGO_TRANSITIONS: dict[EstadoPago, set[EstadoPago]] = {
    EstadoPago.pendiente: {EstadoPago.aprobado, EstadoPago.rechazado},
    EstadoPago.aprobado: set(),  # terminal
    EstadoPago.rechazado: set(),  # terminal
}

SESION_TRANSITIONS: dict[EstadoSesion, set[EstadoSesion]] = {
    EstadoSesion.activa: {EstadoSesion.cerrada, EstadoSesion.expirada},
    EstadoSesion.cerrada: set(),  # terminal
    EstadoSesion.expirada: set(),  # terminal
}

_ALL: dict[str, dict] = {
    "mesa": MESA_TRANSITIONS,
    "pedido": PEDIDO_TRANSITIONS,
    "reserva": RESERVA_TRANSITIONS,
    "pago": PAGO_TRANSITIONS,
    "sesion_mesa": SESION_TRANSITIONS,
}


def validar_transicion(maquina: str, actual, nueva) -> None:
    """Raise ``TransicionInvalidaError`` si ``actual`` → ``nueva`` no está
    declarada en la máquina indicada.

    Args:
        maquina: clave en ``_ALL`` ("mesa", "pedido", "reserva", "pago",
            "sesion_mesa").
        actual: estado origen (Enum).
        nueva: estado destino (Enum).
    """
    transitions = _ALL[maquina]
    permitidas = transitions.get(actual, set())
    if nueva not in permitidas:
        raise TransicionInvalidaError(maquina, actual, nueva)


def puede_transicionar(maquina: str, actual, nueva) -> bool:
    """Versión booleana de :func:`validar_transicion` (sin raise)."""
    transitions = _ALL[maquina]
    return nueva in transitions.get(actual, set())

File: panel_admin/lib/core/state_machines.dart
// core/state_machines.dart — port 1:1 de
// backend/app/core/state_machines.py (ARCHITECTURE.md Pattern 5).
//
// Módulo PURO: no importa Firestore/Auth — solo las tablas de transiciones.
// Las features importan `validarTransicion()` en sus controllers/mutaciones;
// los tests unitarios corren en milisegundos.
//
// Cada tabla mapea `estado_origen -> {estados_destino}` (strings idénticos
// a los enums de Python — Firestore guarda los mismos valores). Un `Set`
// vacío marca un estado TERMINAL (no admite más transiciones).
//
// Diferencia intencional vs Python: los estados son `String` (no Enum)
// porque llegan de/entran a Firestore como strings.

/// Transición de estado no permitida por la máquina de estados del dominio.
///
/// Atributos (port de `TransicionInvalidaError`):
/// * [maquina] — nombre de la máquina ("mesa", "pedido", ...).
/// * [actual] — estado origen.
/// * [nueva] — estado destino rechazado.
class TransicionInvalidaException implements Exception {
  const TransicionInvalidaException(this.maquina, this.actual, this.nueva);

  final String maquina;
  final String actual;
  final String nueva;

  @override
  String toString() =>
      "[$maquina] transición '$actual' → '$nueva' no permitida";
}

// ---- Tablas (diff 1:1 contra state_machines.py, mismo orden/valores) ----

/// Set vacío = terminal (port de MESA_TRANSITIONS).
const Map<String, Set<String>> mesaTransitions = {
  'disponible': {'reservada', 'ocupada'},
  'reservada': {'ocupada', 'disponible'},
  'ocupada': {'limpieza'},
  'limpieza': {'disponible'},
};

/// Port de PEDIDO_TRANSITIONS (rechazado y pagado terminales).
const Map<String, Set<String>> pedidoTransitions = {
  'borrador': {'enviado'},
  'enviado': {'aceptado', 'rechazado'},
  'aceptado': {'en_preparacion'},
  'en_preparacion': {'servido'},
  'servido': {'pagado'},
  'rechazado': <String>{}, // terminal
  'pagado': <String>{}, // terminal
};

/// Port de RESERVA_TRANSITIONS (cancelada terminal).
const Map<String, Set<String>> reservaTransitions = {
  'pendiente': {'confirmada', 'cancelada'},
  'confirmada': {'cancelada'},
  'cancelada': <String>{}, // terminal
};

/// Port de PAGO_TRANSITIONS (aprobado y rechazado terminales).
const Map<String, Set<String>> pagoTransitions = {
  'pendiente': {'aprobado', 'rechazado'},
  'aprobado': <String>{}, // terminal
  'rechazado': <String>{}, // terminal
};

/// Port de SESION_TRANSITIONS (cerrada y expirada terminales).
const Map<String, Set<String>> sesionTransitions = {
  'activa': {'cerrada', 'expirada'},
  'cerrada': <String>{}, // terminal
  'expirada': <String>{}, // terminal
};

/// Registro de las 5 máquinas (port de `_ALL`; clave `sesion_mesa`).
const Map<String, Map<String, Set<String>>> _all = {
  'mesa': mesaTransitions,
  'pedido': pedidoTransitions,
  'reserva': reservaTransitions,
  'pago': pagoTransitions,
  'sesion_mesa': sesionTransitions,
};

/// Tabla de transiciones de una máquina (port del acceso a `_ALL` en los
/// tests Python — permite verificar cobertura sin exponer el registro).
Map<String, Set<String>> transicionesDe(String maquina) {
  final transitions = _all[maquina];
  if (transitions == null) {
    // Python levanta KeyError — aquí ArgumentError (error de programador).
    throw ArgumentError.value(maquina, 'maquina', 'máquina desconocida');
  }
  return transitions;
}

/// Lanza [TransicionInvalidaException] si `actual` → `nueva` no está
/// declarada en la máquina indicada (port de `validar_transicion`).
///
/// Estados sin entrada en la tabla (o terminales) rechazan todo.
void validarTransicion(String maquina, String actual, String nueva) {
  final permitidas = transicionesDe(maquina)[actual] ?? const <String>{};
  if (!permitidas.contains(nueva)) {
    throw TransicionInvalidaException(maquina, actual, nueva);
  }
}

/// Versión booleana de [validarTransicion] (port de `puede_transicionar`).
bool puedeTransicionar(String maquina, String actual, String nueva) {
  return transicionesDe(maquina)[actual]?.contains(nueva) ?? false;
}
